File: Paradigma Funcional/enigmacesar.hs
{-
    ENIGMA DE CESAR
    APLICACOES DE PARADIGMAS DE LINGUAGEM DE PROGRAMACAO - 2017.2 - UFCG
    EQUIPE: ALICE FERNANDES
            ITALO MENEZES
            IVYNA ALVES
            THAIS TOSCANO
            VALTER LUCENA

-}

-- imports
import System.Random
import Charadas
import Textos
import Data.Char
import Palavras
import System.Exit

-- botar aqui as constantes --
listaCharadas = Charadas.charadas
listaFacil = Palavras.facil
listaMedio = Palavras.medio
listaDificil = Palavras.dificil
limiteChave = 4


-- funcao principal
main = do
    Textos.telaInicial
    Textos.desenhaEsfinge
    putStrLn $ "Pressione 1 para as regras, 2 para jogar agora!"
    opcao <- getLine
    if (read opcao) == 1
        then do
            Textos.regras
            putStrLn $ "Digite 1 para jogar agora, ou 0 para sair"
            novaOpcao <- getLine
            if (read novaOpcao) == 1
                then jogar
            else
                return ()
    else
        jogar

jogar = do
    let nivel = 1
    mostraCharada nivel

-- falta inserir a parte de mostrar as palavras e
mostraCharada nivel = do
    if nivel > 3 then return ()
    else do    
        indexCharada <- geraIndice 0 (length listaCharadas - 1)
        chave <- geraIndice 1 limiteChave
        let charada = fst (listaCharadas !! indexCharada)
        let resposta = snd (listaCharadas !! indexCharada)
        let cifrada = cifra charada chave        
        putStrLn $ "\n" ++ cifrada
        putStrLn $ "\nDescifre as palavras abaixo:"
        escolhePalavra 0 cifrada chave 0 nivel
        putStrLn $ "Resposta: "
        respostaUsuario <- getLine
        let resultado = respostaUsuario == resposta
        if resultado
            then mostraCharada (nivel+1)
        else do
            putStrLn $ "Você perdeu :/"
            putStrLn $ "Pressione 2 para jogar novamente ou 0 para sair:"
            opcao <- readLn
            if opcao == 2 then jogar else (exitWith ExitSuccess)

selecionaLista :: Int -> [String]
selecionaLista nivel
    | nivel == 1 = listaFacil
    | nivel == 2 = listaMedio
    | nivel == 3 = listaDificil

--funcao que exibe palavras
escolhePalavra n charada chaveCharada i nivel = do
     if n == 3 then return()
     else do
         let palavras = selecionaLista nivel :: [String]
         indexPalavra <- geraIndice 0 (length palavras - 1)
         chave <- geraIndice 1 limiteChave
         let palavra = palavras !! indexPalavra
         let palavraCifrada = cifra palavra chave
         putStrLn $ show (n+1) ++ ")" ++ palavraCifrada
         putStrLn $ "Dica: chave = " ++ show chave
         putStrLn $ "Resposta: "
         respostaUsuario <- getLine
         if respostaUsuario == palavra then do
             let pedaco = (i+1) * (div (length charada) 3)
             let string = decifraPedaco charada chaveCharada 0 pedaco ++ drop (pedaco+1) charada
             putStrLn string
             escolhePalavra (n + 1) charada chaveCharada (i+1) nivel
         else do
             putStrLn $ "Você perdeu :/"
             putStrLn $ "Pressione 2 para jogar novamente ou 0 para sair:"
             opcao <- readLn
             if opcao == 2 then jogar else (exitWith ExitSuccess)

-- funcao que gera um indice aleatorio    
geraIndice :: Int -> Int -> IO Int
geraIndice inicio limite = randomRIO(inicio, limite)

-- funcoes para nao deixar que os caracteres saiam do escopo [a..z]
letraPNum :: Char -> Int
letraPNum letra = ord letra - ord 'a'

numPLetra :: Int -> Char
numPLetra num = chr (num + ord 'a')

-- funcao para fazer o descolamento de uma letra
desloca :: Char -> Int -> Char
desloca letra desloc
    | (isAlpha letra) = numPLetra (( letraPNum (toLower letra) + desloc) `mod` 26)
    | otherwise = letra

-- cifra de cesar
cifra :: String -> Int -> String
cifra palavra desloc = [desloca palav desloc | palav <- palavra]

-- funcao que retorna uma parte de uma string
substring :: String -> Int -> Int -> String
substring palavra inicio fim = take (fim-inicio+1) (drop (inicio) palavra)
    
-- funcao que decifra apenas uma parte de uma string
decifraPedaco :: String -> Int -> Int -> Int -> String
decifraPedaco palavra chave inicio fim = cifra (substring palavra inicio fim) (0-chave)

File: Paradigma Logico/enigmacesar.pl
/* 	
	ENIGMA DE CESAR
	APLICACOES DE PARADIGMAS DE LINGUAGEM DE PROGRAMACAO - 2017.2 - UFCG
	EQUIPE: ALICE FERNANDES
			ITALO MENEZES
			IVYNA ALVES
			THAIS TOSCANO
			VALTER LUCENA
*/

:- use_module(library(apply)).
:- use_module(library(random)).
:- use_module(library(ctypes)).

:- [textos].
:- [charadas].
:- [palavras].

/* Predicado que criptografa uma string a partir do deslocamento D */
crip(L1, D, L2) :-
    string_to_list(L1, R),
	maplist(cifra(D), R, Y),
	string_to_list(L2, Y).

/* Aplica o deslocamento D */
cifra(D, X, Y) :- string_to_list(C, [X]), (is_alpha(C) -> Y is ((((X - 97) + D) mod 26) + 97) ; Y is X).


/* acessa o elemento associado a um indice */
getElement(0, [E|_], E).
getElement(I, [_|T], R) :- X is I-1, getElement(X, T, R).
    
geraChave(X) :- random(1, 4, X).
geraIndice(Limite, Indice) :- random(0, Limite, Indice).

iniciaJogo() :-
	write('Pressione 1 para as regras ou 2 para jogar.'), nl,
	read(Opcao), nl,
	selecao(Opcao). 

selecao(0) :- write('Voce encerrou o jogo'), nl, halt(0).
selecao(1) :- regras(), nl, write('Pressione 2 para jogar, ou 0 para sair do jogo'), nl, read(Opcao), selecao(Opcao).
selecao(2) :- jogar(1).
selecao(_) :- write('Voce nao leu direito as instrucoes. Reinicie o jogo!'), nl, halt(0).

nivel(1,X) :- facil(X).
nivel(2,X) :- medio(X).
nivel(3,X) :- dificil(X).

jogar(X) :- X > 3, writeln('Digite 2 para jogar novamente ou 0 para encerrar o jogo.'),
			read(Opcao), selecao(Opcao).
			
jogar(Nivel) :- mostraCharada(Nivel,RespostaDaCharada).


mostraCharada(Nivel, _) :- Nivel > 3, writeln('Parabens! Voce finalizou o jogo!'), halt(0).
mostraCharada(Nivel, RespostaDaCharada) :- charadas(D), length(D, Limite),
			geraIndice(Limite, IndiceCharada),
			getElement(IndiceCharada,D,(Charada, RespostaDaCharada)),
			geraChave(Chave),
			crip(Charada, Chave, Criptografada),
			write('Nivel '), write(Nivel), nl, nl,
			writeln(Criptografada),
			mostraPalavra(Nivel, 1, Chave, Charada, RespostaDaCharada).			
			
mostraPalavra(Nivel, Controle, Chave, Charada, RespostaDaCharada) :- Controle > 3, 
			write('Digite resposta: '), 
			read(R1), 
			atom_string(R1, SR),
			atom_string(RespostaDaCharada, SResposta),
			acertouCharada(SResposta, SR),
			NovoNivel is Nivel + 1, 
			mostraCharada(NovoNivel, _).


mostraPalavra(Nivel, Controle, ChaveCharada, Charada, RespostaDaCharada) :- nivel(Nivel,Palavras),
			length(Palavras, Limite),
			geraIndice(Limite, IndicePalavra),
			getElement(IndicePalavra,Palavras,Palavra),
			geraChave(ChavePalavra),
			crip(Palavra, ChavePalavra, PalavraCrip),
			Reverte is (0 - ChavePalavra), crip(PalavraCrip, Reverte, Resp), /*apagar depois, e a resposta v tbm*/
			write(Controle), write(') '), write(PalavraCrip), nl,
			write('Dica: Chave = '), write(ChavePalavra),nl,
			read(R1),
			atom_string(R1, RespostaDaPalavra),
			atom_string(Palavra, SPalavra),
			acertouPalavra(SPalavra, RespostaDaPalavra, Controle, ChaveCharada, Charada),
			NovoControle is Controle + 1,
			mostraPalavra(Nivel, NovoControle, ChaveCharada, Charada, RespostaDaCharada).


result(S1, S2, R) :- atom_string(S1, R1), atom_string(S2, R2), (R1 == R2 -> R = 1; R = 0).

acertouPalavra(S, S, Controle, Chave, Charada) :- 
			crip(Charada, Chave, Criptografada),
			string_length(Charada, Tamanho),
			Pedaco is ceiling(Controle * (Tamanho/3)),
			Resto is (Tamanho - Pedaco),
			sub_string(Charada, 0, Pedaco, Resto, SubCharada),
			sub_string(Criptografada, Pedaco, Resto, 0, SubCrip),
			string_concat(SubCharada, SubCrip, Nova),
			writeln(Nova).
			
acertouPalavra(S, _, _, _, _) :- writeln('Voce perdeu!. Digite 0 para encerrar ou 2 para jogar novamente.'),
			read(Opcao), selecao(Opcao).

acertouCharada(S, S) :- writeln("Parabens!"), nl.
acertouCharada(S, _) :- writeln('Voce perdeu!. Digite 0 para encerrar ou 2 para jogar novamente.'),
			read(Opcao), selecao(Opcao).
			

:- initialization(main).

main :-
	telaInicial(),nl,
	desenhaEsfinge(),nl,
	iniciaJogo(), nl,
	halt(0).
	
	

